### ISSUE TITLE
how to run migrations by code

### ISSUE THREAD

**Issue by andreghisleni** [prisma/prisma#13549]
I have a project with a main database, called general and inside it a table of companies, each company has a different database.
My problem is the following, in this project I use the typeorm, but I am considering migrating to the prisma, the problem is that in the typeorm I can run the migrations inside the connection.

```typescript
await connection.runMigrations({
       transaction: 'all',
});
```

Is there any way to get it to run in the same way as typeorm here in prisma??

I can connect to another db with a custom url:
```ts
import { PrismaClient } from '@prisma/client'

const client1 = new PrismaClient({ datasources: { db: { url: 'postgres://localhost/db1' }} })
```
https://github.com/prisma/prisma/issues/2443#issuecomment-630679118

**Comment by janpio** [comment]
What exactly do you want to do? Run migrations via code on the same connection you use to run queries against your database? Why does `migrate deploy` and similar not work for you?

**Comment by andreghisleni** [comment]
it would be more or less that, the idea is to have several connections by the same file just changing the url, the problem that I would have to run inside the code, that's why this command would not work

**Comment by janpio** [comment]
So you want to run multiple migrations with the same script/code on different databases/connection strings?

**Comment by andreghisleni** [comment]
exactly that, and I didn't find anything in the documentation about it.
do you know if it is possible?

**Comment by janpio** [comment]
Currently that is not easily possible. The best approach here would be to script the usage of `prisma db push` or `prisma migrate deploy` and modifying the environment variable that holds the connection string each time.

**Comment by dongyuwei** [comment]
we do it this way:

import { execSync } from 'child_process';
```
execSync(`pnpx prisma migrate deploy --schema db/shared/shared.prisma`, { stdio: 'inherit' });
```

If your prisma shema need some environment variables, just pass it to the cmd, eg :

```
execSync(
    `export DATABASE_URL=${config.database.url}?schema=${tenantId} && pnpx prisma migrate deploy --schema db/tenants/tenants.prisma`,
    { stdio: 'inherit' },
  );
```

**Comment by andreghisleni** [comment]
> we do it this way:
> 
> ```
> import { execSync } from 'child_process';
> execSync(`pnpx prisma migrate deploy --schema db/shared/shared.prisma`, { stdio: 'inherit' });
> ```
> 
> If your prisma shema need some environment variables, just pass it to the cmd, eg :
> 
> ```
> import { execSync } from 'child_process';
> execSync(
>     `export DATABASE_URL=${config.database.url}?schema=${tenantId} && pnpx prisma migrate deploy --schema db/tenants/tenants.prisma`,
>     { stdio: 'inherit' },
>   );
> ```

Thanks, it wasn't quite what I was looking for but it helped.
is this variable ambente only in the execution of this command or will it be able to be accessed later within the system (linux)?

**Comment by eemelipa** [comment]
Here's also older issue related to having programmatic access to the prisma migrator: https://github.com/prisma/prisma/issues/4703

**Comment by Divaaan** [comment]
@dongyuwei Are you running those commands in a serverless environment by any chance?

I'm trying to run the commands on Azure Functions but getting and error stating
```Error: Command failed: npx prisma migrate deploy --schema ${schemaPath}```

Anyone else tried it on Azure Functions before?

Not sure if there is anything else I should pass through?

**Comment by Divaaan** [comment]
I managed to get a workaround going, instead of uploading the node_modules to my Azure Function, I only upload the build contents and do an `npm i` and `prisma generate` via `spawnSync`/`execSync` once the function initializes. This is the only way I managed to figure it out which is not the greatest way of doing it. Can probably write a script as well to do this for less overhead and function calls.

**Comment by 0gust1** [comment]
@Divaaan we also have this issue, in an 'classic' Azure App Service context. 
It maybe worth to create  a separate issue for that ? (as it's kind of unrelated to the current one).

**Comment by kbariotis** [comment]
@janpio I have a use case for you. We're trying out https://github.com/testcontainers/testcontainers-node and we would like each of our test cases to bring up their own containers (postgres in this case), migrate the full database using Prisma programmatic API which will be fed with the temporal hosts/ports given from testcontainers (because it assigns different ones in each container) and then have all test cases run in parallel.

So programmatic access to Prisma will allow us to do that in just a cleaner way instead of running a script directly from the test case. Thanks, hope that makes sense

**Comment by elijaholmos** [comment]
@kbariotis have you found a workaround in the meantime?

**Comment by moishinetzer** [comment]
> @janpio I have a use case for you. We're trying out https://github.com/testcontainers/testcontainers-node and we would like each of our test cases to bring up their own containers (postgres in this case), migrate the full database using Prisma programmatic API which will be fed with the temporal hosts/ports given from testcontainers (because it assigns different ones in each container) and then have all test cases run in parallel.
> 
> So programmatic access to Prisma will allow us to do that in just a cleaner way instead of running a script directly from the test case. Thanks, hope that makes sense

I have exactly the same use-case, any updates on this ticket?

**Comment by jamesrowe08** [comment]
This would be greatly appreciated as there's no straight forward way when using SQLite, packaging a client application and using an installer.

**Comment by floelhoeffel** [comment]
👋 Hello Everyone!

We are looking into a design for a "Prisma Migrate library". If this sounds relevant to you, please help us with this 5 min survey so we can learn more about your specific needs and priorities.

https://prisma103696.typeform.com/to/btmMPJH6

🙇 Thank you very much!

**Comment by iduuck** [comment]
Thanks @floelhoeffel. I would expect something like described in this issue to be a little like the migrations we have in Ruby on Rails. I want one `up` path, and one `down` path, so that I am free to roll-back and re-deploy the migration.

**Comment by jamesaspence** [comment]
I responded to the survey as well. I am currently trying to set up a test DB in both local and CI, and having that involves (at least locally) defining a separate set of DB credentials. Being able to run the migrations via code would be ideal since I could then define the DB connection in code and run migrations before tests begin. However, right now, attempting to run `yarn migrate dev` or `yarn migrate reset` during test startup will automatically pull in values in my `.env` file and not use my test values

**Comment by lamuertepeluda** [comment]
I am perhaps a bit late to the party here. Although I filled the survey, here's how I run it in my tests using [testcontainers](https://github.com/testcontainers/testcontainers-node) and `ts-jest`. Hope it helps somebody. If you have feedbacks, they're more than welcome.

```ts
import { GenericContainer, StartedTestContainer, Wait } from "testcontainers";
import { promisify } from "node:util";
import { exec } from "node:child_process";
import { getPrismaClient } from "../common/db"; // returns singleton instance as suggested by documentation
import { seedDB } from "../fixtures/seed-db"; // a scripts that fills the database with some data used by the test

const execAsync = promisify(exec);

describe("prisma mysql tests", () => {
  const MYSQL_ROOT_PASSWORD = "somepassword";
  const MYSQL_DATABASE = "loki_accounts";
  const MYSQL_USER = "mysql";
  const MYSQL_PASSWORD = "userpassword";
  let DATABASE_URL: string;

  let container: StartedTestContainer;
  const env = process.env;
  process.env = { ...env };

  // spin and initialize database
  beforeAll(async () => {
    container = await new GenericContainer("mysql:8.0.31")
      .withWaitStrategy(Wait.forHealthCheck())
      .withEnvironment({
        MYSQL_ROOT_PASSWORD,
        MYSQL_DATABASE,
        MYSQL_USER,
        MYSQL_PASSWORD,
      })
      .withHealthCheck({
        test: [
          "CMD-SHELL",
          `mysqladmin ping -h localhost -u root -p${MYSQL_ROOT_PASSWORD}`,
        ],
        interval: 3000,
        timeout: 3000,
        retries: 15,
        startPeriod: 10000,
      })
      .withExposedPorts(3306)
      .start();
    const host = container.getHost();
    const port = container.getMappedPort(3306);
    DATABASE_URL = `mysql://root:${MYSQL_ROOT_PASSWORD}@${host}:${port}/${MYSQL_DATABASE}`;
    process.env = { ...env, DATABASE_URL };
    await execAsync("npx prisma db push", {
      env: process.env,
    });
    await seedDB();
  }, 60000);

  beforeEach(() => {
    jest.resetModules();
    process.env = { ...env, DATABASE_URL };
  });

  afterEach(() => {
    process.env = env;
  });

  // stop database container
  afterAll(async () => {
    await container.stop();
  });

  it("it should find 2 entities", async () => {
    const prisma = getPrismaClient();
    // the seed adds two entity records
    const entities = await prisma.entity.findMany();
    expect(entities.length).toBe(2);
  });

});
```

I also override the DATABASE_URL env var in a way that reflects the temporary container credentials.

I guess this setup could be used for testing migrations as well, although a proper API would help to avoid using process.exec

**Comment by jamesaspence** [comment]
I was likewise using `process.exec` to run migrations programmatically. I ended up giving up and shelving this until such a time as this API is available.

**Comment by yazan73** [comment]
Hell, Any update for this issue?

**Comment by olalonde** [comment]
Duplicate of https://github.com/prisma/prisma/issues/4703 I believe

**Comment by chenyuncai** [comment]
Need this too.

The migrate script need a database_url env, but in most case, the CI building process didnot have any access to know about this. it’s dangerous。Sometimes, the db instance is  a internal k8s pod without any port exosed, the db can only access in the protected network use a service-name（docker-compose or k8s service），In this case we have to run a empty migrate pod before other pod. if we have a "run migdate in code" ability, we can run it in init code or trigger by any restful api or make a rollback from a webpage command, maybe we can show the migrate records in webpage to check the migrate status.

**Comment by fabianwikstrom** [comment]
need this too!

**Comment by fabianwikstrom** [comment]
need this too!

**Comment by basitanwer** [comment]
We definitely need this especially when the database is in an isolated subnet.

**Comment by izilla** [comment]
Need this as well

**Comment by vetlevo** [comment]
We would also really need this feature!

**Comment by Tiketti** [comment]
People, kindly stop posting "me too" messages. They are not adding to the discussions, they're only creating noise for anybody watching the issue.

If you wish to signify that this issue is important to you, use the emoji reactions for the original post or any relevant comments. I'm sure the dev team follows upvotes and can do the math.

**Comment by shtse8** [comment]
Hello,

I'm encountering a specific challenge integrating Prisma with Docker for my production setup. In my production Docker layer, I typically avoid installing Prisma because the project has already been built, which should eliminate the need for Prisma at runtime. However, due to the necessity of running database migrations (prisma db migrate), I find myself required to install Prisma in the production environment regardless.

Additionally, I'm utilizing Nitro (https://nitro.unjs.io/guide/tasks) as my backend, which offers a compelling feature for handling database migrations. However, the lack of support for executing migrations programmatically within Prisma limits my ability to leverage Nitro's full capabilities for this purpose. This setup creates an additional layer of complexity and somewhat undermines the streamlined deployment process I aim for with Docker and Nitro.

Is there a recommended approach or workaround that would allow for the execution of Prisma migrations in a more integrated manner with Nitro, without necessitating a manual step or the installation of Prisma in the production environment? Any guidance or suggestions on how to better streamline this process would be greatly appreciated.

Thank you for your time and assistance.

**Comment by hectorm** [comment]
> Hello,
> 
> I'm encountering a specific challenge integrating Prisma with Docker for my production setup. In my production Docker layer, I typically avoid installing Prisma because the project has already been built, which should eliminate the need for Prisma at runtime. However, due to the necessity of running database migrations (prisma db migrate), I find myself required to install Prisma in the production environment regardless.
> 
> Additionally, I'm utilizing Nitro (https://nitro.unjs.io/guide/tasks) as my backend, which offers a compelling feature for handling database migrations. However, the lack of support for executing migrations programmatically within Prisma limits my ability to leverage Nitro's full capabilities for this purpose. This setup creates an additional layer of complexity and somewhat undermines the streamlined deployment process I aim for with Docker and Nitro.
> 
> Is there a recommended approach or workaround that would allow for the execution of Prisma migrations in a more integrated manner with Nitro, without necessitating a manual step or the installation of Prisma in the production environment? Any guidance or suggestions on how to better streamline this process would be greatly appreciated.
> 
> Thank you for your time and assistance.

I'm exactly in the same boat as you and ended up with this code that uses Prisma's internal API (which has no guarantee of stability).

```typescript
import PrismaInternals from "@prisma/internals";
import PrismaMigrate from "@prisma/migrate";

// [...]

export default defineTask({
  meta: {
    description: "Run database migrations",
  },
  run: async () => {
    let migrate;

    try {
      const schema = await PrismaInternals.getSchemaPath();
      if (!schema) {
        logger.error("No schema found");
        return { result: false };
      }

      migrate = new PrismaMigrate.Migrate(schema);

      const { migrations } = await migrate.listMigrationDirectories();
      if (migrations.length > 0) {
        logger.info(`${migrations.length} migration(s) found`);
      } else {
        logger.info("No migration found");
      }

      const { appliedMigrationNames } = await migrate.applyMigrations();
      if (appliedMigrationNames.length > 0) {
        logger.info(`The following migration(s) have been applied: ${appliedMigrationNames.join(", ")}`);
      } else {
        logger.info("No pending migrations to apply");
      }

      return { result: true };
    } catch (error) {
      logger.error(error, "Error running migrations");
      return { result: false, error };
    } finally {
      migrate?.stop();
    }
  },
});
```

You will also need to modify the `nuxt.config.ts` file to copy the migrations and schema engine.

```typescript
import fs from "node:fs/promises";
import path from "node:path";

// [...]

export default defineNuxtConfig({
  // [...]
  hooks: {
    "nitro:build:public-assets": async (nitro) => {
      // Copy Prisma schema and migrations to the output directory
      const prismaInDir = path.join(__dirname, "prisma");
      const prismaOutDir = path.join(nitro.options.output.dir, "prisma");
      await fs.cp(prismaInDir, prismaOutDir, { recursive: true });

      // Copy Prisma schema engine to the output directory
      const prismaEnginesInDir = path.join(__dirname, "node_modules", "@prisma", "engines");
      const prismaEnginesOutDir = path.join(nitro.options.output.serverDir, "node_modules", "@prisma", "engines");
      const prismaEnginesFiles = await fs.readdir(prismaEnginesInDir);
      for (const file of prismaEnginesFiles) {
        if (/^schema-engine-.+$/.test(file)) {
          await fs.cp(path.join(prismaEnginesInDir, file), path.join(prismaEnginesOutDir, file));
        }
      }
    },
  },
  // [...]
});
```

I have not been able to test it extensively, but so far I have not found any problems. But it would be ideal if Prisma would provide a stable API for this.

**Comment by quinn** [comment]
I was able to get this to work by doing this:

```ts
import { MigrateDeploy } from '@prisma/migrate'
const migrate = MigrateDeploy.new()
await migrate.parse([])
```

it seems the migration code is mixed directly in with the command parsing/cli code, unfortunately. Also, this package is for internal use only, so this may suddenly break for me in the future. Another also, this can't be passed an existing prisma connection config, so will rely on prisma's internal semantics for resolving a connection to the db (e.g. `$DATABASE_URL`)

**Comment by janpio** [comment]
No. Currently this is not scheduled or being worked on. We will post here when we do that.

**Comment by lekhasy** [comment]
It is not possible to use [pgmock](https://github.com/stack-auth/pgmock) without this feature.
pgmock is in-memory version of postgres and it mock TCP connection to the db - which only work if you connect to it inside nodejs.
If we use execSync, it will not work since the migration will run on a different process.

**Comment by heldercostaa** [comment]
This worked for my tests run executing in a `beforeEach` without a big impact, if anyone is interested:
```typescript
// reset-db.ts
import { PrismaClient } from "@prisma/client";
import { execSync } from "node:child_process";

async function resetDb() {
  await prisma.$transaction([
    prisma.table.deleteMany(),
    // ...
  ]);
}

export default async () => {
  try {
    await resetDb();
  } catch (err) {
    execSync(`npx prisma migrate deploy`, { stdio: "inherit" });
    await resetDb();
  }
};
```
```typescript
// mytest.spec.ts
describe("", () =>{
  beforeEach(async () => {
    await resetDb();
  });

  // it("should....
}
```

**Comment by Jobinjose01** [comment]
@heldercostaa Thanks its working
```
async function runPrismaMigrations() {
    try {
        // Capture the output of the command
        const statusOutput = execSync('npx prisma migrate status').toString();
        console.log(statusOutput);

        if (statusOutput.includes('There are 1 or more migrations that have not yet been applied')) {
            const deployOutput = execSync('npx prisma migrate deploy').toString();
            console.log(deployOutput);
            logger.info(deployOutput);
            console.log("Migrations applied successfully.");
        } else {
            console.log("No pending migrations.");
        }
    } catch (error) {
        console.error("Error applying migrations:", error);
        logger.error(error);
    }
}
```

**Comment by JerryBels** [comment]
I wanted to run tests on a specific, clean DB each time, with seeder done. I installed dotenvx and created scripts in package.json:

```
"scripts": {
    "prisma:generate:dev": "dotenvx run -f .env.development -- npx prisma generate",
    "prisma:generate:test": "dotenvx run -f .env.testing -- npx prisma generate",
    "prisma:migrate:dev": "dotenvx run -f .env.development -- npx prisma migrate deploy",
    "prisma:migrate:test": "dotenvx run -f .env.testing -- npx prisma migrate deploy",
    "prisma:push:dev": "dotenvx run -f .env.development -- prisma db push",
    "prisma:push:test": "dotenvx run -f .env.testing -- prisma db push",
    "prisma:seed:dev": "dotenvx run -f .env.development -- tsx prisma/seed.ts",
    "prisma:seed:test": "dotenvx run -f .env.testing -- tsx prisma/seed.ts",
    "prisma:reset:dev": "dotenvx run -f .env.development -- npx prisma migrate reset --force",
    "prisma:reset:test": "dotenvx run -f .env.testing -- npx prisma migrate reset --force"
  },
  "prisma": {
    "seed": "tsx prisma/seed.ts"
  },
```

Then in the `setup.ts` I used:

```
beforeAll(async () => {
  // ... more code
  execSync('npm run prisma:migrate:test');
});

beforeEach(async () => {
  // run prisma seeder
  execSync('npm run prisma:reset:test');
});
```

Works great for me.

**Comment by ziazon** [comment]
Here's how I ended up doing it.  mind you, I hate this code, but works for now.  mind you this is from a nestjs perspective.

```typescript
import child_process from 'node:child_process';

import { ConsoleLogger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

export async function runMigration(
  config: ConfigService,
  logger: ConsoleLogger
) {
  const db_url = `postgresql://${config.get('DB_USER')}:${config.get('DB_PASSWORD')}@${config.get('DB_HOST')}:${config.get('DB_PORT')}/${config.get('DB_NAME')}`;

  try {
    await new Promise((resolve, reject) => {
      const process = child_process.exec(
        `export DATABASE_URL=${db_url} && pnpm prisma migrate deploy --schema=./src/prisma/schema.prisma`
      );

      process.stdout.on('data', (chunk) => logger.log(chunk));

      process.on('close', (code) =>
        code != 0 ? reject(process.stderr) : resolve(process.stdout)
      );
    });
  } catch (error) {
    logger.error(error);
  }
}

```

then in my app startup entry point file `main.ts` I have this:

```typescript
  if (config.get('RUN_MIGRATIONS')) await runMigration(config, logger);

  await app.listen(config.get('API_PORT'));

```

Honestly I really hope either a public API is made available (stable one) for managing migrations programmatically, in particular running them on a connection you specify.

**Comment by mitchellston** [comment]
For anybody needing this. I have created a package that can do this, it isn't completed, but it might be interesting for some:
https://www.npmjs.com/package/prisma-migrations-runner?activeTab=readme

(currently only mysql works and rollbacks are not possible via the package)

**Comment by primemb** [comment]
Any updates on this? I understand automatic migrations can be tricky for SQL databases, but for MongoDB it would be incredibly useful if Prisma could apply code changes to the database automatically—just like Mongoose—reducing DevOps overhead when developers don’t have direct database access.

**Comment by delebash** [comment]
After looking through the tests directory I have figured out a few things that make running the cli commands in code work.  Check out https://github.com/prisma/prisma/tree/main/packages/migrate/src/__tests__  for good info.

Here is how you can run commands, check the tests directory above for any specific command parameters.

Note:  this is not use the prisma.config.ts but just the regular shcema.prisma setup, I am just passing it in manually, you could read from the file to get the config, or there is probably a method to extract the config from your schema.prisma

import packages just the ones you need.  I just imported all just to list them.  I am using es modules vs commonjs so we have to import it differently


```

const {
    Migrate,
    DbCommand,
    DbExecute,
    DbPull,
    DbPush,
    // DbDrop,
    DbSeed,
    getDatabaseVersionSafe,
    MigrateCommand,
    MigrateDeploy,
    MigrateDev,
    MigrateDiff,
    MigrateReset,
    MigrateResolve,
    MigrateStatus,
} = pkg

import pkg from '@prisma/migrate';
import fs from 'fs';

//Define you connection same as defined in your schema.prisma file you
const schemaBase = `
generator client {
provider = "prisma-client-js"
output= "./prisma/generated/prisma-client" 
previewFeatures = ["driverAdapters", "queryCompiler"]
 }
datasource db {
  provider = "postgres"
  url      = "postgres://postgres:test@127.0.0.1:5432/test"
}
  `

//Run the primsa db pull command
const introspect = new DbPull()
const result = introspect.parse(['--schema=./prisma/schema.prisma'], schemaBase)
  
```

If you want to get the output of the introspection before you write it to file you can.  Note with this way we do not even need anything defined in the schema.prisma file we can create dynamic connections this way.

```
const migrate = await Migrate.setup({})
const engine = migrate.engine
//Generate introspection schema without writing to file
const introspectionResult = await engine.introspect({
    schema: {
        files: [{path: 'schema.prisma', content: schemaBase}],
    },
    baseDirectoryPath: process.cwd(),
    viewsDirectoryPath: '',
})
console.log(introspectionResult.schema.files[0])
//Write introspection schema to file
await fs.writeFileSync('./prisma/schema.prisma', introspectionResult.schema.files[0].content)
```

```
//Push
const result = DbPush.new().parse([],schemaBase)
```

**Comment by delebash** [comment]
Anyone want to try and get the generate client cli from code working?

**Comment by delebash** [comment]
Also some prisma code uses an npm module called execa https://www.npmjs.com/package/execa  Makes it easy to call commands

```
import {execa, ExecaError} from 'execa';

try {
	await execa`npx prisma generate`;
} catch (error) {
	if (error instanceof ExecaError) {
		console.log(error);
	
```}

**Comment by delebash** [comment]
Also a package to run migrations in code with additional features  https://www.npmjs.com/package/prisma-migrations

**Comment by diced** [comment]
It's been over ~6 years since this issue and other issues with the same idea were opened. I haven't necessarily looked into all the comments here, but for what it's worth I don't think I'm missing out on much lol. I'm just wondering about the state of having "programatic access" to the migrations API which I have been using in a pretty janky way, but it works for my use case. Unfortunately, a few months ago some Prisma 6.x update made it considerably annoying and harder to use with, making me have to pin my Prisma versions to `6.13.0`, while the current version as of writing this comment is `7.5.0`.

What I want to do is simple, and mostly revolves around the ease of use for my users:
- I want to run migrations automatically on app start up
- I don't want to continue using `prisma@6.13.0`

Reasons like this make it easier to choose something like drizzle-orm over Prisma, which yes, could be done, but I am a small team of developers (literally just me!) and I do not have the time to spend migrating my codebase that heavily relies on Prisma to use something different like drizzle-orm (which is considerably easier to use programmatically). I have tried this and it's a way better DX, but I can't put the time commitment into migrating my entire codebase.

If you want to see my implementation of running migrations automatically see the files I linked below:
https://github.com/diced/zipline/blob/trunk/src/lib/db/migration/index.ts
What this file does is basically what the `prisma migrate deploy` command does (I looked at the source of that command, and made it work in my case). In my main file, I call `runMigrations()` which does everything seamlessly in the background.

With all that said, I'm coming back to my first point: what's the state?
